src/utils/packetIdAllocator.ts

```diff
@@ -0,0 +1,79 @@
+/* Keeps track of free and used packet ids by storing available IDs as intervals */
+export class PacketIdAllocator {
+  /**
+   * Intervals are stored in reverse order since push/pop from end of a list is
+   * faster than shift/unshift from beginning.
+   */
+  private _intervals: [number, number][] = [[65535, 1]];
+
+  async allocate(): Promise<number> {
+    if (this._intervals.length === 0) {
+      /* TODO: return promise that resolves when a new packet ID is available */
+      throw new Error('No more packet IDs available');
+    }
+
+    const lastInterval = this._intervals[this._intervals.length - 1] as [number, number];
```

Comment:
is there a reason you are typecasting here? Why not define the type of lastInterval as [number, number]

Reply:
Typescript makes this `[number, number] | undefined` because it's not aware that list is non-empty at the point that we check it. We have a check earlier for the list being empty earlier so it's safe to assume that `list[list.length - 1]` will always be not undefined